c_src/membrane_vpx_plugin/vpx_decoder.c

```diff
@@ -0,0 +1,150 @@
+  const int number_of_planes = (img->fmt == VPX_IMG_FMT_NV12) ? 2 : 3;
+
+  size_t image_size = 0;
+
+  for (int plane = 0; plane < number_of_planes; ++plane) {
+    Dimensions plane_dimensions = get_plane_dimensions(img, plane);
+    image_size +=
+        plane_dimensions.width * plane_dimensions.height * bytes_per_pixel;
+  }
+  return image_size;
+}
+
+void get_output_frame_from_image(const vpx_image_t *img,
+                                 UnifexPayload *output_frame) {
+  const int bytes_per_pixel = (img->fmt & VPX_IMG_FMT_HIGHBITDEPTH) ? 2 : 1;
+
+  // Assuming that for nv12 we write all chroma data at once
+  const int number_of_planes = (img->fmt == VPX_IMG_FMT_NV12) ? 2 : 3;
+  unsigned char *frame_data = output_frame->data;
+
+  for (int plane = 0; plane < number_of_planes; ++plane) {
+    const unsigned char *buf = img->planes[plane];
+    const int stride = img->stride[plane];
+    Dimensions plane_dimensions = get_plane_dimensions(img, plane);
+
+    for (unsigned int y = 0; y < plane_dimensions.height; ++y) {
+      size_t bytes_to_write = bytes_per_pixel * plane_dimensions.width;
+      memcpy(frame_data, buf, bytes_to_write);
+      buf += stride;
+      frame_data += bytes_to_write;
+    }
+  }
+}
+
+void alloc_output_frame(UnifexEnv *env, const vpx_image_t *img,
+                        UnifexPayload **output_frame) {
+  *output_frame = unifex_alloc(sizeof(UnifexPayload));
+  unifex_payload_alloc(env, UNIFEX_PAYLOAD_BINARY, get_image_byte_size(img),
+                       *output_frame);
+}
+
+PixelFormat get_pixel_format_from_image(vpx_image_t *img) {
+  switch (img->fmt) {
+  case VPX_IMG_FMT_I422:
+    return PIXEL_FORMAT_I422;
+
+  case VPX_IMG_FMT_I420:
+    return PIXEL_FORMAT_I420;
+
+  case VPX_IMG_FMT_I444:
+    return PIXEL_FORMAT_I444;
+
+  case VPX_IMG_FMT_YV12:
+    return PIXEL_FORMAT_YV12;
+
+  case VPX_IMG_FMT_NV12:
+    return PIXEL_FORMAT_NV12;
+
+  default:
+    return PIXEL_FORMAT_I420;
+  }
+}
+
+UNIFEX_TERM decode_frame(UnifexEnv *env, UnifexPayload *frame, State *state) {
+  vpx_codec_iter_t iter = NULL;
+  vpx_image_t *img = NULL;
+  unsigned int frames_cnt = 0, max_frames = 2;
+  PixelFormat pixel_format = PIXEL_FORMAT_I420;
+  UnifexPayload **output_frames =
+      unifex_alloc(max_frames * sizeof(*output_frames));
+
+  if (vpx_codec_decode(&state->codec_context, frame->data, frame->size, NULL,
+                       0)) {
+    return decode_frame_result_error(env, "Decoding frame failed");
+  }
+
+  while ((img = vpx_codec_get_frame(&state->codec_context, &iter)) != NULL) {
+    if (frames_cnt >= max_frames) {
+      max_frames *= 2;
+      output_frames =
+          unifex_realloc(output_frames, max_frames * sizeof(*output_frames));
+    }
+
+    alloc_output_frame(env, img, &output_frames[frames_cnt]);
+    get_output_frame_from_image(img, output_frames[frames_cnt]);
+    pixel_format = get_pixel_format_from_image(img);
+    frames_cnt++;
+  }
```

Comment:
Why do we need to update `max_frames` ? Could you add some comment? :D

Reply:
the idea is to double the amount of allocated frames once we hit frame count equal to it, i guess `allocated_frames` would be more appropriate than `max_frames`

Reply:
good point, now I understand